### PR TITLE
cop: allow light tasks to run when the semaphore has no permits available

### LINKED ISSUE #6544: Remove the semaphore for tidb_query
tidb_query execution in batch is changed to use coroutine, there is risk that big requests that keep coming will cause OOM and starvation. So a semaphore is used to limit requests in execution in TiKV.

However, if all resources defined by the semaphore is occupied by big requests, small requests are denied. Under this circumstance, we still cannot benefit from the new thread pool and coroutined coprocessor execution.

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

Previously, we use a semaphore to limit concurrently running tasks. The intention is to avoid OOM and starvation due to too many big requests, but it may block small tasks from running.

In this PR, we just **try to acquire** a permit from the semaphore. It can fail, but when it fails, the execution must be quick done. If the execution runs for too long (longer than 5ms), it is cancelled and forced to acquire a permit from the semaphore.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Refer to a related PR or issue link (optional)

This PR intends to resolve #6544, but still using the semaphore

###  Benchmark result if necessary (optional)

I run sysbench read_only with 64 threads, and from at about 45s,, I launch a lot (256 threads) of big coprocessor requests.

Before: 

```
[ 10s ] thds: 64 tps: 2734.76 qps: 16420.64 (r/w/o: 10945.12/0.00/5475.51) lat (ms,99%): 30.26 err/s: 0.00 reconn/s: 0.00            
[ 20s ] thds: 64 tps: 3807.92 qps: 22850.35 (r/w/o: 15234.90/0.00/7615.45) lat (ms,99%): 27.17 err/s: 0.00 reconn/s: 0.00            
[ 30s ] thds: 64 tps: 3811.59 qps: 22871.07 (r/w/o: 15247.48/0.00/7623.59) lat (ms,99%): 27.17 err/s: 0.00 reconn/s: 0.00            
[ 40s ] thds: 64 tps: 3806.84 qps: 22838.55 (r/w/o: 15224.47/0.00/7614.08) lat (ms,99%): 27.66 err/s: 0.00 reconn/s: 0.00            
[ 50s ] thds: 64 tps: 1861.87 qps: 11171.20 (r/w/o: 7447.47/0.00/3723.73) lat (ms,99%): 27.66 err/s: 0.00 reconn/s: 0.00             
[ 60s ] thds: 64 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,99%): 0.00 err/s: 0.00 reconn/s: 0.00                           
[ 70s ] thds: 64 tps: 0.40 qps: 2.90 (r/w/o: 2.10/0.00/0.80) lat (ms,99%): 24994.79 err/s: 0.00 reconn/s: 0.00                       
[ 80s ] thds: 64 tps: 0.60 qps: 5.50 (r/w/o: 4.30/0.00/1.20) lat (ms,99%): 27846.48 err/s: 0.00 reconn/s: 0.00                       
[ 90s ] thds: 64 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,99%): 0.00 err/s: 0.00 reconn/s: 0.00
```

This PR:

```
[ 10s ] thds: 64 tps: 3766.66 qps: 22619.15 (r/w/o: 15079.74/0.00/7539.42) lat (ms,99%): 28.67 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 64 tps: 3810.64 qps: 22863.66 (r/w/o: 15242.47/0.00/7621.19) lat (ms,99%): 27.17 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 64 tps: 3762.71 qps: 22575.77 (r/w/o: 15050.25/0.00/7525.52) lat (ms,99%): 29.19 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 64 tps: 3805.73 qps: 22834.87 (r/w/o: 15223.51/0.00/7611.36) lat (ms,99%): 27.66 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 64 tps: 3793.25 qps: 22757.29 (r/w/o: 15170.69/0.00/7586.60) lat (ms,99%): 28.16 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 64 tps: 3774.69 qps: 22649.56 (r/w/o: 15100.37/0.00/7549.19) lat (ms,99%): 28.16 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 64 tps: 3808.79 qps: 22851.57 (r/w/o: 15233.58/0.00/7617.99) lat (ms,99%): 27.17 err/s: 0.00 reconn/s: 0.00
[ 80s ] thds: 64 tps: 3773.08 qps: 22637.58 (r/w/o: 15091.82/0.00/7545.76) lat (ms,99%): 28.16 err/s: 0.00 reconn/s: 0.00
[ 90s ] thds: 64 tps: 3811.91 qps: 22872.46 (r/w/o: 15248.74/0.00/7623.72) lat (ms,99%): 27.66 err/s: 0.00 reconn/s: 0.00
[ 100s ] thds: 64 tps: 3667.61 qps: 22006.55 (r/w/o: 14670.93/0.00/7335.62) lat (ms,99%): 42.61 err/s: 0.00 reconn/s: 0.00
[ 110s ] thds: 64 tps: 3792.01 qps: 22749.18 (r/w/o: 15165.05/0.00/7584.13) lat (ms,99%): 27.66 err/s: 0.00 reconn/s: 0.00
[ 120s ] thds: 64 tps: 3785.37 qps: 22715.52 (r/w/o: 15144.98/0.00/7570.54) lat (ms,99%): 28.67 err/s: 0.00 reconn/s: 0.00
[ 130s ] thds: 64 tps: 3769.42 qps: 22616.45 (r/w/o: 15077.40/0.00/7539.05) lat (ms,99%): 28.67 err/s: 0.00 reconn/s: 0.00
```

### CODE REVIEW

File: src/coprocessor/metrics.rs [hunk 1]
```diff
@@ -67,6 +68,26 @@ lazy_static! {
         "Total bytes of response body"
     )
     .unwrap();
+    pub static ref COPR_ACQUIRE_SEMAPHORE_RESULT: CoprAcquireSemaphoreResultCounterVec =
+        register_static_int_counter_vec!(
+            CoprAcquireSemaphoreResultCounterVec,
+            "tikv_coprocessor_acquire_semaphore_result",
```
Comment (andylokandy) [inline src/coprocessor/metrics.rs]: Could give it a name like `tikv_coprocessor_acquire_semaphore_type`.

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -1324,4 +1374,80 @@ mod tests {
...
+
+        // semaphore has free permit, we only handle once
+        let resp = cop
+            .handle_unary_request(ReqContext::default_for_test(), handler_builder_gen())
+            .wait()
+            .unwrap();
+        assert_eq!(resp.data, 1usize.to_le_bytes());
+
+        // remove all permits
+        std::mem::forget(cop.semaphore.as_ref().unwrap().try_acquire().unwrap());
```
[the remark this answers is not in the document]
Comment (sticnarf) [reply src/coprocessor/endpoint.rs]: Fixed.

File: src/coprocessor/metrics.rs (hunk 1 above)
Comment (sticnarf) [reply src/coprocessor/metrics.rs]: I hesitated which to use...If you like I am willing to use `_type`.

**Pull request merged**